test/e2e/framework/create.go

```diff
@@ -0,0 +1,528 @@
+// - the latest stable API version for each entity is used, regardless of what
+//   is specified in the manifest files
+func (f *Framework) CreateFromManifest(patch func(object interface{}), files ...string) (func(), error) {
+	var destructors []func() error
+	var cleanupHandle CleanupActionHandle
+	cleanup := func() {
+		if cleanupHandle == nil {
+			// Already done.
+			return
+		}
+		RemoveCleanupAction(cleanupHandle)
+
+		// TODO (?): use same logic as framework.go for determining
+		// whether we are expected to clean up? This would change the
+		// meaning of the -delete-namespace and -delete-namespace-on-failure
+		// command line flags, because they would also start to apply
+		// to non-namespaced entities.
+		for _, destructor := range destructors {
+			if err := destructor(); err != nil && !apierrs.IsNotFound(err) {
+				Logf("deleting failed: %s", err)
+			}
+		}
+	}
+	cleanupHandle = AddCleanupAction(cleanup)
+
+	err := visitManifests(func(data []byte) error {
+		// Ignore any additional fields for now, just determine what we have.
+		var what What
+		if err := runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), data, &what); err != nil {
+			return errors.Wrap(err, "decode TypeMeta")
+		}
+
+		factory := Factories[what]
+		if factory == nil {
+			return errors.Errorf("item of type %+v not supported", what)
+		}
+
+		item := factory.New()
+		object := item.ToObject()
+		if err := runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), data, object); err != nil {
+			return errors.Wrapf(err, "decode %+v", what)
+		}
+
+		f.PatchItem(object)
+		if patch != nil {
+			patch(object)
+		}
+
+		destructor, err := item.Create(f)
+		if destructor != nil {
+			destructors = append(destructors, destructor)
+		}
+		return err
+	}, files...)
+
+	return cleanup, err
+}
+
+// What is a subset of metav1.TypeMeta which (in contrast to
+// metav1.TypeMeta itself) satisfies the runtime.Object interface.
+type What struct {
+	Kind string `json:"kind"`
+}
+
+func (in *What) DeepCopy() *What {
+	return &What{Kind: in.Kind}
+}
+
+func (in *What) DeepCopyInto(out *What) {
+	out.Kind = in.Kind
+}
+
+func (in *What) DeepCopyObject() runtime.Object {
+	return &What{Kind: in.Kind}
+}
+
+func (in *What) GetObjectKind() schema.ObjectKind {
+	return nil
+}
+
+// ItemFactory provides support for creating one particular entity.
+// The type gets exported because other packages might want to
+// extend the set of pre-defined factories.
+type ItemFactory interface {
+	// New returns a new empty entity.
+	New() Item
+}
+
+// Item contains one particular entity instance and provides
+// the method for creating it.
+type Item interface {
+	// ToObject grants access to the entity.
+	ToObject() runtime.Object
+	// Create is responsible for creating the entity after
+	// populating and patching it. It returns an error or
+	// a cleanup function for the created entity.
+	Create(f *Framework) (func() error, error)
+}
+
+var Factories = map[What]ItemFactory{
```

Comment:
This model becomes challenging if we ever have to support CRDs in the future.

Reply:
Can you elaborate further what the problem would be? I'm not familiar with CRDs, but looking at https://github.com/kubernetes/csi-api/blob/master/pkg/crd/crd.go and https://github.com/kubernetes/csi-api/blob/master/pkg/crd/testdata/csidriver.yaml, a CRD looks like normal API objects that can be populate using this approach from a .yaml file.

Reply:
We don't have to use this approach for CRDs, or we do the patching inside the test which knowns about the CRD. I'm pretty sure that some solution can be found.